### ISSUE TITLE
Protocol abort at high rate

### ISSUE THREAD

**Issue by dariol83** [dariol83/ccsds#5]
Attached the logs from the ReatMetric system:

INFO: sagr=abc.spack=xyz.fsl-fg=1.cltu=cltu01: State transition from READY to START_PENDING
ago 08, 2020 7:27:41 PM eu.dariolucia.ccsds.sle.utl.si.ServiceInstance setServiceInstanceState
INFO: sagr=abc.spack=xyz.fsl-fg=1.cltu=cltu01: State transition from START_PENDING to ACTIVE
ago 08, 2020 7:27:41 PM eu.dariolucia.ccsds.sle.utl.si.cltu.CltuServiceInstance handleCltuScheduleStatusReportReturn
INFO: sagr=abc.spack=xyz.fsl-fg=1.cltu=cltu01: Schedule status report return received, positive result
ago 08, 2020 7:28:36 PM eu.dariolucia.ccsds.sle.utl.network.tml.TmlChannel protocolErrorDetected
**SEVERE: Protocol error detected on channel TmlChannelClient[192.168.0.129:23233]@1579782619 with reason UNKNOWN_TYPE_ID, header=[-53, -9, 29, 0, 32, 0, -80, 39]**
ago 08, 2020 7:28:36 PM eu.dariolucia.ccsds.sle.utl.si.ServiceInstance setServiceInstanceState
INFO: sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91: State transition from ACTIVE to UNBOUND
ago 08, 2020 7:28:36 PM eu.dariolucia.reatmetric.driver.spacecraft.sle.RafServiceInstanceManager handleOperation
**WARNING: sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91: Discarding unhandled RAF operation: eu.dariolucia.ccsds.sle.generated.ccsds.sle.transfer.service.raf.outgoing.pdus.RafTransferBuffer
ago 08, 2020 7:28:36 PM eu.dariolucia.reatmetric.driver.spacecraft.sle.RafServiceInstanceManager handleOperation
WARNING: sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91: Discarding unhandled RAF operation: eu.dariolucia.ccsds.sle.generated.ccsds.sle.transfer.service.raf.outgoing.pdus.RafTransferBuffer**
ago 08, 2020 7:28:36 PM eu.dariolucia.ccsds.sle.utl.network.tml.TmlChannel sendHbtMessage
WARNING: Cannot send HBT on channel TmlChannelClient[192.168.0.129:23233]@1579782619, disconnected
ago 08, 2020 7:28:36 PM eu.dariolucia.reatmetric.driver.spacecraft.sle.SleServiceInstanceManager onPduDecodingError
SEVERE: sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91: SLE PDU decoding error
ago 08, 2020 7:30:29 PM eu.dariolucia.reatmetric.api.transport.AbstractTransportConnector disconnect
INFO: Transport connector sagr=abc.spack=xyz.fsl-fg=1.cltu=cltu01 disconnecting
ago 08, 2020 7:30:29 PM eu.dariolucia.ccsds.sle.utl.si.ServiceInstance setServiceInstanceState
INFO: sagr=abc.spack=xyz.fsl-fg=1.cltu=cltu01: State transition from ACTIVE to STOP_PENDING
ago 08, 2020 7:30:29 PM eu.dariolucia.ccsds.sle.utl.si.ServiceInstance setServiceInstanceState
INFO: sagr=abc.spack=xyz.fsl-fg=1.cltu=cltu01: State transition from STOP_PENDING to READY
ago 08, 2020 7:30:29 PM eu.dariolucia.ccsds.sle.utl.si.ServiceInstance setServiceInstanceState

**Comment by slux83** [comment]
Hi,

I just saw the same issue here while debugging. I blocked the tm packet processor with a break-point and somehow disrupted the processing at SLE level.

```
...
Service instance sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91 status updated: ACTIVE
Service instance sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91 status updated: UNBOUND
sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91: Unexpected disconnection detected: UNKNOWN_TYPE_ID
sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91: Discarding unhandled RAF operation: eu.dariolucia.ccsds.sle.generated.ccsds.sle.transfer.service.raf.outgoing.pdus.RafTransferBuffer
Service instance sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91 status updated: UNBOUND
sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91: Discarding unhandled RAF operation: eu.dariolucia.ccsds.sle.generated.ccsds.sle.transfer.service.raf.outgoing.pdus.RafTransferBuffer
Service instance sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91 status updated: UNBOUND
sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91: Discarding unhandled RAF operation: eu.dariolucia.ccsds.sle.generated.ccsds.sle.transfer.service.raf.outgoing.pdus.RafTransferBuffer
Service instance sagr=abc.spack=xyz.rsl-fg=1.raf=onlc91 status updated: UNBOUND
...
```

**Comment by dariol83** [comment]
The main reason was related to the size of the buffering within the API (1000 operations) causing a write session to easily go in timeout. The size of the buffers has been reduced and a backpressure test implemented.

**Comment by slux83** [comment]
I'd like to reopen this issue because the previous fix did not completely resolve the problem. I managed to reproduce the issue by applying a lot of back pressure and I think the main problem was that the class `eu.dariolucia.ccsds.sle.utl/src/main/java/eu/dariolucia/ccsds/sle/utl/network/tml/TmlChannel.java` does not completely protect the output stream of the socket from concurrent write access. The attached patch seems to solve the problem (applied on the branch 0.3.0). 
[tml_concurrency_fix.patch.txt](https://github.com/dariol83/ccsds/files/5851690/tml_concurrency_fix.patch.txt)

**Comment by dariol83** [comment]
Thank you for the patch, indeed there is a concurrency issue in writing to the socket output stream between the calling thread that sends PDUs, and the HBT thread. I implemented the suggested fix, but I did not re-use the currently defined lock: if a write becomes blocking, a lock on the lock object would prevent a shutdown of the SI and the execution of a termination. I introduced a separate lock just to protect the write() operation.
